project/npda/tasks.py

```diff
@@ -0,0 +1,352 @@
+# Python Imports
+import collections
+from collections import defaultdict
+from datetime import datetime
+import json
+import logging
+
+# Django Imports
+from django.apps import apps
+from django.db import transaction
+
+# Third party imports
+from celery import shared_task
+import numpy as np
+import pandas as pd
+
+# Project Imports
+from .models import Visit
+from .forms.patient_form import PatientForm
+from .forms.visit_form import VisitForm
+from .general_functions.csv.progress_recorder import ProgressTracker
+
+# Logging setup
+logger = logging.getLogger(__name__)
+
+
+@shared_task
+def hello():
+    """
+    THIS IS A SCHEDULED TASK THAT IS CALLED AT 06:00 EVERY DAY
+    THE CRON DATE/FREQUENCY IS SET IN SETTING.PY
+    """
+    logger.debug("0600 cron check task ran successfully")
+
+
+@shared_task(bind=True)
+def save_patient_and_visits_to_submission(
+    self, patient_row_json, patient_dict, patient_group_dict, pz_code, submission_id
+):
+    """
+    Accepts a list of patient visits, creates a patient instance (validating using a patient form), then iterates
+    through the visits, creating a form for each and gathering up the errors
+    """
+
+    """
+    Helper functions
+    """
+
+    def retain_errors_and_invalid_field_data(form):
+        # We want to retain fields even if they're invalid so that we can return them to the user
+        # Use the field value from cleaned_data, falling back to data if it's not there
+        for key, value in form.cleaned_data.items():
+            setattr(form.instance, key, value)
+
+        for key, value in form.data.items():
+            if key not in form.cleaned_data:
+                setattr(form.instance, key, value)
+
+        form.instance.is_valid = form.is_valid()
+        form.instance.errors = (
+            None if form.is_valid() else form.errors.get_json_data(escape_html=True)
+        )
+
+    def record_errors_from_form(errors_to_return, row_index, form):
+        for field, errors in form.errors.as_data().items():
+            for error in errors:
+                errors_to_return[row_index][field].extend(error.messages)
+
+    def row_to_dict(row, model, csv_headings):
+        ret = {}
+        for entry in csv_headings:
+            if "model" in entry and apps.get_model("npda", entry["model"]) == model:
+                model_field_name = entry["model_field"]
+                model_field_definition = model._meta.get_field(model_field_name)
+
+                csv_value = row[entry["heading"]]
+                model_field_value = csv_value_to_model_value(
+                    model_field_definition, csv_value
+                )
+
+                if model_field_name in [
```

Comment:
My gut feeling is that we would do this in `read_csv` instead? Ideally whatever dataframe comes out of that can be processed without additional cleaning